src/lib.rs

```diff
@@ -1132,40 +1130,237 @@ pub unsafe trait TryFromBytes {
+    ///
+    /// // The only valid value of this type is the bytes `0xC0C0`.
+    /// #[derive(TryFromBytes, KnownLayout, Immutable)]
+    /// #[repr(C, packed)]
+    /// struct C0C0(C0, C0);
+    ///
+    /// #[derive(TryFromBytes, KnownLayout, Immutable)]
+    /// #[repr(C)]
+    /// struct Packet {
+    ///     magic_number: C0C0,
+    ///     mug_size: u8,
+    ///     temperature: u8,
+    ///     marshmallows: [[u8; 2]],
+    /// }
+    ///
+    /// let bytes = &[0xC0, 0xC0, 240, 77, 0, 1, 2, 3, 4, 5][..];
+    ///
+    /// let packet = Packet::try_ref_from(bytes).unwrap();
+    ///
+    /// assert_eq!(packet.mug_size, 240);
+    /// assert_eq!(packet.temperature, 77);
+    /// assert_eq!(packet.marshmallows, [[0, 1], [2, 3], [4, 5]]);
+    ///
+    /// // These bytes are not valid instance of `Packet`.
+    /// let bytes = &[0x70, 0xC0, 240, 77, 0, 1, 2, 3, 4, 5][..];
+    /// assert!(Packet::try_ref_from_prefix(bytes).is_err());
+    /// ```
     #[must_use = "has no side effects"]
     #[inline]
-    fn try_ref_from(bytes: &[u8]) -> Result<&Self, TryCastError<&[u8], Self>>
+    fn try_ref_from(candidate: &[u8]) -> Result<&Self, TryCastError<&[u8], Self>>
     where
         Self: KnownLayout + Immutable,
     {
         util::assert_dst_is_not_zst::<Self>();
-        let candidate = Ptr::from_ref(bytes)
-            .try_cast_into_no_leftover::<Self>()
-            .map_err(|e| e.map_src(|src| src.as_ref()).into())?;
-
-        // This call may panic. If that happens, it doesn't cause any soundness
-        // issues, as we have not generated any invalid state which we need to
-        // fix before returning.
-        //
-        // Note that one panic or post-monomorphization error condition is
-        // calling `try_into_valid` (and thus `is_bit_valid`) with a shared
-        // pointer when `Self: !Immutable`. Since `Self: Immutable`, this panic
-        // condition will not happen.
-        let candidate = candidate.try_into_valid().map_err(|e| e.with_src(bytes).into());
+        match Ptr::from_ref(candidate).try_cast_into_no_leftover::<Self>() {
+            Ok(candidate) => {
+                // This call may panic. If that happens, it doesn't cause any soundness
+                // issues, as we have not generated any invalid state which we need to
+                // fix before returning.
+                //
+                // Note that one panic or post-monomorphization error condition is
+                // calling `try_into_valid` (and thus `is_bit_valid`) with a shared
+                // pointer when `Self: !Immutable`. Since `Self: Immutable`, this panic
+                // condition will not happen.
+                match candidate.try_into_valid() {
+                    Ok(valid) => Ok(valid.as_ref()),
+                    Err(e) => Err(e.map_src(|src| src.as_bytes().as_ref()).into()),
+                }
+            }
+            Err(e) => Err(e.map_src(Ptr::as_ref).into()),
+        }
```

Comment:
Rewritten in this manner for consistency with our other `from_ref` methods, and because doing so allows us to name `try_ref_from`'s argument "candidate" without thinking of new, weird identifier names.